sympy/core/relational.py

```diff
@@ -1138,3 +1049,204 @@ def _eval_relation(cls, lhs, rhs):
     '<': StrictLessThan,
     'lt': StrictLessThan,
 }
+
+
+def _n2(a, b):
+    """Return (a - b).evalf(2) if a and b are comparable, else None.
+    This should only be used when a and b are already sympified.
+    """
+    # /!\ it is very important (see issue 8245) not to
+    # use a re-evaluated number in the calculation of dif
+    if a.is_comparable and b.is_comparable:
+        dif = (a - b).evalf(2)
+        if dif.is_comparable:
+            return dif
+
+
+@dispatch(Expr, Expr)
+def _eval_is_le(lhs, rhs):
+    return None
+
+
+@dispatch(Expr, Expr)
+def _eval_is_ge(lhs, rhs):
+    return None
+
+
+@dispatch(Basic, Basic)
+def _eval_is_eq(lhs, rhs):
+    return None
+
+
+@dispatch(Tuple, Expr)
+def _eval_is_eq(lhs, rhs):# noqa:F811
+    return False
+
+
+@dispatch(Tuple, AppliedUndef)
+def _eval_is_eq(lhs, rhs):# noqa:F811
+    return None
+
+
+@dispatch(Tuple, Symbol)
+def _eval_is_eq(lhs, rhs):# noqa:F811
+    return None
+
+
+@dispatch(Tuple, Tuple)
+def _eval_is_eq(lhs, rhs): # noqa:F811
+    if len(lhs) != len(rhs):
+        return False
+
+    return fuzzy_and(fuzzy_bool(is_eq(s, o)) for s, o in zip(lhs, rhs))
+
+
+def is_lt(lhs, rhs):
+    return fuzzy_not(is_ge(lhs, rhs))
+
+
+def is_gt(lhs, rhs):
+    return fuzzy_not(is_le(lhs, rhs))
+
+
+def is_le(lhs, rhs):
+    return is_ge(rhs, lhs)
+
+
+def is_ge(lhs, rhs):
+    if not (isinstance(lhs, Expr) and isinstance(rhs, Expr)):
+        raise TypeError("Can only compare inequalities with Expr")
+
+    retval = _eval_is_ge(lhs, rhs)
+
+    if retval is not None:
+        return retval
+    else:
+        n2 = _n2(lhs, rhs)
+        if n2 is not None:
+            # use float comparison for infinity.
+            # otherwise get stuck in infinite recursion
+            if n2 in (S.Infinity, S.NegativeInfinity):
+                n2 = float(n2)
+            return _sympify(n2 >= 0)
+        if lhs.is_extended_real and rhs.is_extended_real:
+            if (lhs.is_infinite and lhs.is_extended_positive) \
+                     or (rhs.is_infinite and rhs.is_extended_negative):
+                return True
+            diff = lhs - rhs
+            if diff is not S.NaN:
+                rv = diff.is_extended_nonnegative
+                if rv is not None:
+                   return rv
+
+
+def is_neq(lhs, rhs):
+    return fuzzy_not(is_eq(lhs, rhs))
+
+
+def is_eq(lhs, rhs):
+    from sympy.core.add import Add
+    from sympy.functions.elementary.complexes import arg
+    from sympy.simplify.simplify import clear_coefficients
+    from sympy.utilities.iterables import sift
+
+
+    retval = _eval_is_eq(lhs, rhs)
+    if retval is not None:
+        return retval
+
+    if dispatch(type(lhs), type(rhs)) != dispatch(type(rhs), type(lhs)):
+        retval = _eval_is_eq(rhs, lhs)
+        if retval is not None:
+            return retval
+
+    retval = lhs._eval_Eq(rhs)
+    if retval is not None:
```

Comment:
Is there still an `_eval_Eq` method defined on `Basic`? Maybe it needs to still be there for backward compatibility...

I would want to do `getattr(lhs, '_eval_Eq', None)` to allow the possibility that the input doesn't have an `_eval_Eq` method but I guess that isn't needed if it's defined on `Basic`.

There should be a comment here explaining that `_eval_Eq` is only being called for backwards compatibility reasons and that anyone wanting to overload this should use `_eval_is_eq`.